### PR TITLE
Allow max request size to be user-specified

### PR DESCRIPTION
**Pull request by jefferai**
This turned out to be way more impactful than I'd expected because I
felt like the right granularity was per-listener, since an org may want
to treat external clients differently from internal clients. It's pretty
straightforward though.

This also introduces actually using request contexts for values, which
so far we have not done (using our own logical.Request struct instead),
but this allows non-logical methods to still get this benefit.

### CODE REVIEW

**Comment by jefferai** [comment]
Pretty trivially tested (the acutal limiting logic is the same, just the plumbing):

Limiting to smaller than default:
```
$ vault kv put -address=http://127.0.0.1:8200 secret/foo k=@CHANGELOG.md
Key              Value
---              -----
created_time     2018-06-24T16:27:25.632026852Z
deletion_time    n/a
destroyed        false
version          1

$ vault kv put -address=http://127.0.0.1:8202 secret/foo k=@CHANGELOG.md
Error writing data to secret/data/foo: Error making API request.

URL: PUT http://127.0.0.1:8202/v1/secret/data/foo
Code: 413. Errors:

* failed to parse JSON input: http: request body too large

$ truncate --size=21000 CHANGELOG.md 

$ vault kv put -address=http://127.0.0.1:8202 secret/foo k=@CHANGELOG.md
Key              Value
---              -----
created_time     2018-06-24T16:28:44.109742609Z
deletion_time    n/a
destroyed        false
version          2
```
Unlimited (ignore that I accidentally used `vault write`, doesn't matter):
```
$ vault write -address=http://127.0.0.1:8200 secret/foo k=@bin/vault
Error writing data to secret/foo: Error making API request.

URL: PUT http://127.0.0.1:8200/v1/secret/foo
Code: 413. Errors:

* failed to parse JSON input: http: request body too large

$ vault write -address=http://127.0.0.1:8202 secret/foo k=@bin/vault
Error writing data to secret/foo: Error making API request.

URL: PUT http://127.0.0.1:8202/v1/secret/foo
Code: 404. Errors:


WARNING! The following warnings were returned from Vault:

  * Invalid path for a versioned K/V secrets engine. See the API docs for the
  appropriate API endpoints to use. If using the Vault CLI, use 'vault kv put'
  for this operation.
```

File: helper/forwarding/util.go [hunk 1]
```diff
@@ -56,11 +58,31 @@ func GenerateForwardedHTTPRequest(req *http.Request, addr string) (*http.Request
...
+		max, ok := maxRequestSize.(int64)
+		if !ok {
+			return nil, errors.New("could not parse max_request_size from request context")
+		}
+		if max > 0 {
+			reader = io.LimitReader(req.Body, max)
+		}
+	}
+
+	buf := bytes.NewBuffer(nil)
```
Comment (kalafut) [inline helper/forwarding/util.go]: Could use` ioutil.ReadAll()` here since you want `[]byte` later.

Comment (jefferai) [reply helper/forwarding/util.go]: That's fair, I just kept that part of the logic that was there. Feel free to pull down the branch and change it yourself if you like, too!

Comment (kalafut) [reply helper/forwarding/util.go]: Done!

**Comment by jefferai** [comment]
Thanks!

File: command/server.go [hunk 2]
```diff
@@ -792,7 +808,9 @@ CLUSTER_SYNTHESIS_COMPLETE:
 	// This needs to happen before we first unseal, so before we trigger dev
 	// mode if it's set
 	core.SetClusterListenerAddrs(clusterAddrs)
-	core.SetClusterHandler(vaulthttp.Handler(core))
+	core.SetClusterHandler(vaulthttp.Handler(&vault.HandlerProperties{
+		Core: core,
```
Comment (vishalnayak) [inline command/server.go]: Won't we ever need a size limit on the cluster connection?

File: http/handler.go [hunk 3]
```diff
@@ -326,8 +339,19 @@ func (fs *UIAssetWrapper) Open(name string) (http.File, error) {
...
+	reader := r.Body
+	ctx := r.Context()
+	maxRequestSize := ctx.Value("max_request_size")
+	if maxRequestSize != nil {
+		max, ok := maxRequestSize.(int64)
+		if !ok {
+			return errors.New("could not parse max_request_size from request context")
+		}
+		if max > 0 {
+			reader = http.MaxBytesReader(w, r.Body, max)
```
Comment (vishalnayak) [inline http/handler.go]: Out of curiosity, why do we use `io.LimitReader` above and `http.MaxBytesReader` here?

File: vault/request_handling.go [hunk 4]
```diff
@@ -26,6 +26,13 @@ const (
 	replTimeout = 10 * time.Second
 )
 
+// HanlderProperties is used to seed configuration into a vaulthttp.Handler.
+// It's in this package to avoid a circular dependency
+type HandlerProperties struct {
```
Comment (vishalnayak) [inline vault/request_handling.go]: [Nitpick] Can we call this `HandlerOptions` instead, unless there was a reason to call this `HandlerProperties`?

File: http/handler.go (hunk 3 above)
Comment (jefferai) [reply http/handler.go]: Different function signatures, based on what's available.

File: command/server.go (hunk 2 above)
Comment (jefferai) [reply command/server.go]: By the time we get to the HTTP handler we've already handshaked the cluster connection so we already know if the connection is authorized -- if so we shouldn't limit it.

File: vault/request_handling.go (hunk 4 above)
Comment (jefferai) [reply vault/request_handling.go]: I called it Properties since it's not just user-set options (e.g. Core).

**Pull request merged**